src/sql-parser/src/parser.rs

```diff
@@ -1667,13 +1667,41 @@ impl<'a> Parser<'a> {
                 // FROM S3 BUCKET '<bucket>' OBJECTS FROM SCAN MATCHING '<pattern>'
                 self.expect_keyword(BUCKET)?;
                 let bucket = self.parse_literal_string()?;
-                self.expect_keywords(&[OBJECTS, FROM, SCAN])?;
+                self.expect_keywords(&[OBJECTS, FROM])?;
+                let mut key_sources = Vec::new();
+                let mut parsed_scan = false;
+                while let Some(keyword) = self.parse_one_of_keywords(&[SCAN]) {
+                    match keyword {
+                        SCAN => {
+                            if parsed_scan {
+                                return Err(ParserError::new(
+                                    self.peek_prev_pos(),
+                                    "SCAN may only be specified once",
+                                ));
+                            }
+                            parsed_scan = true;
```

Comment:
This check is a semantic concern, not a syntactic concern. Could you move it to the `sql` crate, per https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#encapsulation?

Reply:
sure.